pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,987 @@
+			case watch.Added, watch.Modified:
+				klog.V(2).InfoS("Secret added or modified", "Secret", klog.KObj(secretMeta))
+				func() {
+					c.bgpPeerPasswordsMutex.Lock()
+					defer c.bgpPeerPasswordsMutex.Unlock()
+
+					secretObj := event.Object.(*corev1.Secret)
+					c.bgpPeerPasswords = make(map[string]string)
+					for k, v := range secretObj.Data {
+						c.bgpPeerPasswords[k] = string(v)
+					}
+				}()
+				c.queue.Add(dummyKey)
+			case watch.Deleted:
+				klog.V(2).InfoS("Secret deleted", "Secret", klog.KObj(secretMeta))
+				func() {
+					c.bgpPeerPasswordsMutex.Lock()
+					defer c.bgpPeerPasswordsMutex.Unlock()
+
+					// Clear the passwords since the Secret is deleted
+					c.bgpPeerPasswords = make(map[string]string)
+				}()
+				c.queue.Add(dummyKey)
+			case watch.Error:
+				klog.V(2).InfoS("Received an error event", "Secret", klog.KObj(secretMeta))
+			}
+		case <-stopCh:
+			return nil
+		}
+	}
+}
+
+func (c *Controller) Run(ctx context.Context, stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting", "controllerName", controllerName)
+	defer klog.InfoS("Shutting down", "controllerName", controllerName)
+
+	cacheSyncs := []cache.InformerSynced{
+		c.nodeListerSynced,
+		c.serviceListerSynced,
+		c.bgpPolicyListerSynced,
+		c.endpointSliceListerSynced,
+	}
+	if c.egressEnabled {
+		cacheSyncs = append(cacheSyncs, c.egressListerSynced)
+	}
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	go wait.NonSlidingUntil(func() {
+		if err := c.watchSecretChanges(stopCh); err != nil {
+			klog.ErrorS(err, "Watch Secret error", "Secret", types.BGPPolicySecretName)
+		}
+	}, time.Second*10, stopCh)
+	go wait.Until(func() {
+		c.worker(ctx)
+	}, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) worker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	_, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(dummyKey)
+
+	if err := c.syncBGPPolicy(ctx); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until another change happens.
+		c.queue.Forget(dummyKey)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		c.queue.AddRateLimited(dummyKey)
+		klog.ErrorS(err, "Syncing BGPPolicy failed, requeue")
+	}
+	return true
+}
+
+func (c *Controller) getEffectiveBGPPolicy() *v1alpha1.BGPPolicy {
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	var oldestBP *v1alpha1.BGPPolicy
+	for _, bp := range allBPs {
+		if c.matchedCurrentNode(bp) {
+			if oldestBP == nil || bp.CreationTimestamp.Before(&oldestBP.CreationTimestamp) {
+				oldestBP = bp
+			}
+		}
+	}
+	return oldestBP
+}
+
+func (c *Controller) syncBGPPolicy(ctx context.Context) error {
+	startTime := time.Now()
+	defer func() {
+		klog.InfoS("Finished syncing BGPPolicy", "durationTime", time.Since(startTime))
+	}()
+
```

Comment:
Regarding a previous conversation about wrapping the context object and using a timeout, I don't know how long syncing the BGPPolicy typically takes. I would recommend using a 30s or 60s timeout for the entire `syncBGPPolicy` call just in case something gets stuck:
```go
func (c *Controller) syncBGPPolicy(ctx context.Context) error {
	ctx, cancel := contextWithTimeoutCause(ctx, 60*time.Second, fmt.Errorf("BGPPolicy took too long to sync"))
	defer cancel()
	// ...
```

Reply:
Added